components/director/internal/tenantmapping/mapper_for_user.go

```diff
@@ -6,55 +6,110 @@ import (
 	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
 )
 
-func NewMapperForUser(staticUserRepo StaticUserRepository, tenantRepo TenantRepository) *mapperForUser {
+func NewMapperForUser(staticUserRepo StaticUserRepository, staticGroupRepo StaticGroupRepository, tenantRepo TenantRepository) *mapperForUser {
 	return &mapperForUser{
-		staticUserRepo: staticUserRepo,
-		tenantRepo:     tenantRepo,
+		staticUserRepo:  staticUserRepo,
+		staticGroupRepo: staticGroupRepo,
+		tenantRepo:      tenantRepo,
 	}
 }
 
 type mapperForUser struct {
-	staticUserRepo StaticUserRepository
-	tenantRepo     TenantRepository
+	staticUserRepo  StaticUserRepository
+	staticGroupRepo StaticGroupRepository
+	tenantRepo      TenantRepository
+}
+
+func stringInSlice(a string, list []string) bool {
+	for _, b := range list {
+		if b == a {
+			return true
+		}
+	}
+	return false
+}
+
+const tenantGroupPrefix = "tenantID="
+
+// GetGroupScopesAndTenant get all scopes from group array, without duplicates
+func GetGroupScopesAndTenant(groups []StaticGroup) (string, string) {
+	var filteredScopesArray []string
+	var tenantID string
+
+	for _, group := range groups {
+		if strings.HasPrefix(group.GroupName, tenantGroupPrefix) {
+			// group name is tenantID
+			tenantID = strings.Replace(group.GroupName, tenantGroupPrefix, "", 1)
+		} else {
+			// group name is
+			for _, scope := range group.Scopes {
+
+				if !stringInSlice(scope, filteredScopesArray) {
+					filteredScopesArray = append(filteredScopesArray, scope)
+				}
+			}
+		}
+	}
+
+	return strings.Join(filteredScopesArray, " "), tenantID
 }
 
 func (m *mapperForUser) GetObjectContext(ctx context.Context, reqData ReqData, username string) (ObjectContext, error) {
 	var externalTenantID, scopes string
+	var staticUser StaticUser
+	var staticGroups []StaticGroup
 
-	staticUser, err := m.staticUserRepo.Get(username)
-	if err != nil {
-		return ObjectContext{}, errors.Wrap(err, fmt.Sprintf("while searching for a static user with username %s", username))
-	}
+	userGroups, err := reqData.GetGroups()
+	log.Infof("GetGroups returned %s\n", userGroups)
 
-	scopes, err = reqData.GetScopes()
 	if err != nil {
-		if !apperrors.IsKeyDoesNotExist(err) {
-			return ObjectContext{}, errors.Wrap(err, "while fetching scopes")
-		}
+		return ObjectContext{}, errors.Wrap(err, fmt.Sprintf("while getting groups for a static user with username %s", username))
+	}
 
-		scopes = strings.Join(staticUser.Scopes, " ")
+	if len(userGroups) > 0 {
+		staticGroups = m.staticGroupRepo.Get(userGroups)
 	}
 
-	externalTenantID, err = reqData.GetExternalTenantID()
-	if err != nil {
-		if !apperrors.IsKeyDoesNotExist(err) {
-			return ObjectContext{}, errors.Wrap(err, "while fetching external tenant")
+	if len(staticGroups) > 0 {
+		// proceed with group scopes flow
+		scopes, externalTenantID = GetGroupScopesAndTenant(staticGroups)
+		log.Infof("Decided to use groups copes %s\n", scopes)
+	} else {
```

Comment:
What if the user is a static user but also sends you groups that are not assigned to any scopes?

Reply:
That's a nice thing to discuss! 🗣

Reply:
We are taking either group or user. So if someone is logging with static user (but also has some random group without scopes), we are treating him as static group user without scopes